compiler/rustc_mir_transform/src/elaborate_box_derefs.rs

```diff
@@ -0,0 +1,80 @@
+//! This pass transforms derefs of Box into a deref of the pointer inside Box
+//! Codegen does not allow box to be directly dereferenced
+
+use crate::MirPass;
+use rustc_hir::def_id::DefId;
+use rustc_index::vec::Idx;
+use rustc_middle::mir::visit::MutVisitor;
+use rustc_middle::mir::*;
+use rustc_middle::ty::subst::Subst;
+use rustc_middle::ty::{Ty, TyCtxt};
+use rustc_target::abi::VariantIdx;
+
+struct ElaborateBoxDerefVistor<'tcx, 'a> {
+    tcx: TyCtxt<'tcx>,
+    unique_did: DefId,
+    local_decls: &'a mut LocalDecls<'tcx>,
+}
+
+impl<'tcx, 'a> MutVisitor<'tcx> for ElaborateBoxDerefVistor<'tcx, 'a> {
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn visit_place(
+        &mut self,
+        place: &mut Place<'tcx>,
+        context: visit::PlaceContext,
+        location: Location,
+    ) {
+        let tcx = self.tcx;
+
+        let mk_place_elems = |boxed_ty: Ty<'tcx>| {
+            let unique_substs = tcx.intern_substs(&[boxed_ty.into()]);
+            let unique_ty = tcx.type_of(self.unique_did).subst(tcx, unique_substs);
+            let ptr_ty = tcx.mk_imm_ptr(boxed_ty);
+
+            [
+                PlaceElem::Field(Field::new(0), unique_ty),
+                PlaceElem::Field(Field::new(0), ptr_ty),
+                PlaceElem::Deref,
+            ]
+        };
+
+        let mut new_projection = Vec::new();
+
+        for (base, proj) in place.iter_projections() {
+            let base_ty = base.ty(self.local_decls, tcx).ty;
+
+            if base_ty.is_box() && proj == PlaceElem::Deref {
+                new_projection.extend(mk_place_elems(base_ty.boxed_ty()));
+            } else {
+                new_projection.push(proj);
```

Comment:
this unconditionally allocates and re-interns all projections, even those that don't have box derefs within them at all.

Reply:
after https://github.com/rust-lang/rust/pull/95649 and its follow up PRs that @ouz-a  is working on right now, this should become much simpler, as we stop having derefs in the middle of things and only have it as the last projection on things, never in the middle.

Reply:
It looks like that pass only operates on `StatementKind::Assign` and that derefs in other places wouldn't be changed. Is this going to be addressed in one of those follow up PRs or am I missing something?

Reply:
Yea, there's a follow up that looks almost ready: https://github.com/ouz-a/rust/tree/terminator_butcher